### PR TITLE
Fix potential issues and bugs in hack/lib/*.sh scripts using shellcheck

### PR DESCRIPTION
**Pull request by ipuustin**
**What this PR does / why we need it**:

This PR is continuing the work for cleaning up the shell scripts using shellcheck and manual inspection. The plan is to make the scripts more robust in case for unexpected input and also to fix potential bugs.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

The changes are for scripts in `hack/lib/` directory, and as library scripts are used from multiple places. It is not trivial to see all possible places from where the scripts are accessed, so careful review is needed. I tried to make sure that the changes would be compatible for older bash versions too.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: hack/lib/etcd.sh [hunk 1]
```diff
@@ -102,7 +102,7 @@ kube::etcd::cleanup() {
 
 kube::etcd::install() {
   (
-    cd "${KUBE_ROOT}/third_party"
+    cd "${KUBE_ROOT}/third_party" || return
```
Comment (stevekuznetsov) [inline hack/lib/etcd.sh]: won't `set -o errexit` accomplish the same thing for us?

File: hack/lib/golang.sh [hunk 2]
```diff
@@ -393,7 +401,7 @@ kube::golang::place_bins() {
     # The substitution on platform_src below will replace all slashes with
     # underscores.  It'll transform darwin/amd64 -> darwin_amd64.
     local platform_src="/${platform//\//_}"
-    if [[ $platform == $host_platform ]]; then
+    if [[ $platform == "$host_platform" ]]; then
```
Comment (stevekuznetsov) [inline hack/lib/golang.sh]: why quote one but not both?

File: hack/lib/util.sh [hunk 3]
```diff
@@ -209,7 +209,7 @@ kube::util::gen-docs() {
 # Puts a placeholder for every generated doc. This makes the link checker work.
 kube::util::set-placeholder-gen-docs() {
   local list_file="${KUBE_ROOT}/docs/.generated_docs"
-  if [ -e ${list_file} ]; then
+  if [ -e "${list_file}" ]; then
```
Comment (stevekuznetsov) [inline hack/lib/util.sh]: could also move to `[[` here

File: hack/lib/util.sh [hunk 4]
```diff
@@ -244,18 +244,18 @@ kube::util::remove-gen-docs() {
 kube::util::group-version-to-pkg-path() {
   staging_apis=(
   $(
-    pushd ${KUBE_ROOT}/staging/src/k8s.io/api > /dev/null
-      find . -name types.go | xargs -n1 dirname | sed "s|\./||g" | sort
-    popd > /dev/null
-  )
-  )
+    cd ${KUBE_ROOT}/staging/src/k8s.io/api &&
```
Comment (stevekuznetsov) [inline hack/lib/util.sh]: might as well quote the arg to `cd` while touching this

File: hack/lib/util.sh [hunk 5]
```diff
@@ -244,18 +244,18 @@ kube::util::remove-gen-docs() {
...
+    find . -name types.go -exec dirname {} \; | sed "s|\./||g" | sort
+  ))
 
   local group_version="$1"
 
-  if [[ " ${staging_apis[@]} " =~ " ${group_version/.*k8s.io/} " ]]; then
-    echo "vendor/k8s.io/api/${group_version/.*k8s.io/}"
-    return
-  fi
+  for api in "${staging_apis[@]:-}"; do
```
Comment (stevekuznetsov) [inline hack/lib/util.sh]: this changes the semantics of the function -- is it correct?

File: hack/lib/util.sh [hunk 6]
```diff
@@ -461,7 +455,12 @@ kube::util::ensure_godep_version() {
 kube::util::ensure_no_staging_repos_in_gopath() {
   kube::util::ensure_single_dir_gopath
   local error=0
-  for repo in $(ls ${KUBE_ROOT}/staging/src/k8s.io); do
+  for repo_file in ${KUBE_ROOT}/staging/src/k8s.io/*; do
```
Comment (stevekuznetsov) [inline hack/lib/util.sh]: quote the `"${KUBE_ROOT}"` bit before the glob?

File: hack/lib/util.sh [hunk 7]
```diff
@@ -461,7 +455,12 @@ kube::util::ensure_godep_version() {
...
+    if [[ ! -d "$repo_file" ]]; then
+      # not a directory or there were no files
+      continue;
+    fi
+    repo=$(basename $repo_file)
```
Comment (stevekuznetsov) [inline hack/lib/util.sh]: quote subshell?

**Comment by fejta** [comment]
/uncc
/assign @stevekuznetsov 
/approve
/hold

Steve feel free to unhold once everything looks good!

File: hack/lib/golang.sh (hunk 2 above)
Comment (ipuustin) [reply hack/lib/golang.sh]: That was pattern matching... The right side of the `==` operator is considered a pattern and is matched, and the quoting just forces it to be matched as a string.

Consider this script:

    #!/bin/bash
  
    bar="foobar"
    foo="foo*"
    
    [[ $bar == $foo ]] && echo "first true"
    [[ "$bar" == "$foo" ]] && echo "second true"
    [[ $bar == "$foo" ]] && echo "third true"
    [[ "$bar" == $foo ]] && echo "fourth true"

I get the output:

    first true
    fourth true

I think the normal use is to typically just quote the right hand side, but quoting the left hand side will not hurt. I'll do that and also improve the commit msg with the example.

File: hack/lib/util.sh (hunk 3 above)
Comment (ipuustin) [reply hack/lib/util.sh]: Good point, I'll do that.

File: hack/lib/util.sh (hunk 6 above)
Comment (ipuustin) [reply hack/lib/util.sh]: Ok.

File: hack/lib/util.sh (hunk 7 above)
Comment (ipuustin) [reply hack/lib/util.sh]: Argh, that's right. I'll do that.

File: hack/lib/util.sh (hunk 4 above)
Comment (ipuustin) [reply hack/lib/util.sh]: Ok. I think a long-term goal is to be able to build kubernetes, when the source tree is in a directory with a space in it's path (such as `/tmp/foo bar/kubernetes`. Right now it doesn't build for many reasons.

File: hack/lib/util.sh (hunk 5 above)
Comment (ipuustin) [reply hack/lib/util.sh]: The semantics changes, yes. The original issue in the code is not that severe -- it requires specifically-crafted code to get a false match from the regexp, even though the array is concatenated to a single string. This is bit complex function and I'm not fully certain the first check is supposed to only handle the `apiname/version` case. Let's drop this patch just to be on the safe side.

File: hack/lib/etcd.sh (hunk 1 above)
Comment (ipuustin) [reply hack/lib/etcd.sh]: It will... and it's set from `init.sh`, which sources this. Good catch. I'll remove the patch.

**Comment by stevekuznetsov** [comment]
Looks reasonable. Might be good in the future to do a PR per commit, so it's quick to review and less risky to merge (as coverage for these is poor in tests).

/lgtm
/hold cancel

**Comment by ipuustin** [comment]
Thanks for the review @stevekuznetsov ! I'll do shorter (even one-commit) PRs in the future, or at least try to group the commits around a common fix (say fixing a single function in a script).

**Pull request merged**